### PR TITLE
ARROW-14100: [C++] subtract(duration, duration) -> duration kernel

### PR DESCRIPTION
**Pull request by rok**
[ARROW-14100](https://issues.apache.org/jira/browse/ARROW-14100)

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 1]
```diff
@@ -976,6 +976,23 @@ TEST_F(ScalarTemporalTest, TestTemporalDifference) {
...
+                      ArrayFromJSON(unit, times_s),
+                      ArrayFromJSON(unit, seconds_between_time));
+  }
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      NotImplemented, testing::HasSubstr("no kernel matching input types"),
+      Subtract(ArrayFromJSON(duration(TimeUnit::SECOND), times_s),
+               ArrayFromJSON(duration(TimeUnit::MILLI), times_s)));
+}
+
```
Comment (edponce) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:995]: Add tests with valid values.

**Review by edponce — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 1 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:995]: Do you mean [like these](https://github.com/apache/arrow/blob/b3cff8f82fd43d3cbbed841f40bb8a3ca0e274e1/cpp/src/arrow/compute/kernels/scalar_temporal_test.cc#L982-L988)?

I'd like to enable cases like `subtract(duration[ms], duration[us]) -> duration[us]`. I'll ping when it's done.

Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1008]: Ok, this now covers cases like `subtract(duration[ms], duration[us]) -> duration[us]`.

**Comment by rok** [comment]
@pitrou could you also take a look at this and #12215?

**Comment by rok** [comment]
Sorry for the push. Had to rebase.

**Comment by rok** [comment]
> Ha, there are conflicts again :-)

It's the overhead price of distributed review chunks :)

**Comment by rok** [comment]
@pitrou this and #12215 should be stable now.

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 2]
```diff
@@ -1242,6 +1242,21 @@ TEST_F(ScalarTemporalTest, TestTemporalSubtractTimeAndDuration) {
...
+      CheckScalarBinary(op, ArrayFromJSON(unit, times_s2), ArrayFromJSON(unit, times_s),
+                        ArrayFromJSON(unit, seconds_between_time));
+    }
+
+    auto seconds_3 = ArrayFromJSON(duration(TimeUnit::SECOND), R"([3, null])");
+    auto milliseconds_2k = ArrayFromJSON(duration(TimeUnit::MILLI), R"([2000, null])");
+    auto milliseconds_1k = ArrayFromJSON(duration(TimeUnit::MILLI), R"([1000, null])");
+    CheckScalarBinary(op, seconds_3, milliseconds_2k, milliseconds_1k);
+  }
+}
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1258]: Can you add tests for overflowing inputs?

Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1284]: Added. However it seems as it duration is not treated as `int64_t` [here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/int_util_internal.h#L59) and overflow is not detected?

Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1284]: Never mind. I was using the wrong operator.

**Review by pitrou — APPROVED** [review]
+1, thank you. I'll just wait for the AppVeyor build to run again.

**Comment by rok** [comment]
Thanks for the review @pitrou. The appveyor issue seems related to the substrait parser. Maybe a rebase will help.

**Pull request closed without merging**